src/frontend/explain_assert.nit

```diff
@@ -0,0 +1,253 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Explain failed assert to the console by modifying the AST.
+#
+# This module implements the service `AAssertExpr::explain_assert_str`,
+# which should be used by the engines.
+#
+# Example assert:
+#
+# ~~~nitish
+# var x = 0.0
+# var y = 1.0
+# assert x.is_approx(y, 0.5)
+# ~~~
+#
+# Produces the following output on failure:
+#
+# ~~~raw
+# Runtime assert: 0.0.is_approx(1.0, 0.5)
+# ~~~
+module explain_assert
+
+import astbuilder
+intrude import literal # for value=
+intrude import typing # for mtype=
+
+import explain_assert_api
+
+redef class ToolContext
+
+	# Phase modifying the AST to explain assets when they fail
+	var explain_assert_phase: Phase = new ExplainAssertPhase(self, [modelize_class_phase, typing_phase, literal_phase])
+end
+
+private class ExplainAssertPhase
+	super Phase
+
+	redef fun process_nmodule(nmodule)
+	do
+		var mmodule = nmodule.mmodule
+		if mmodule == null then return
+
+		# Skip if `mmodule` doesn't have access to `String`
+		var string_class = toolcontext.modelbuilder.try_get_mclass_by_name(nmodule, mmodule, "String")
+		if string_class == null then return
+
+		# Launch a visitor on all elements of the AST
+		var visitor = new UsefulAssertVisitor(toolcontext, mmodule, string_class.mclass_type)
+		visitor.enter_visit nmodule
+	end
+end
+
+# Visitor to find and explain asserts
+private class UsefulAssertVisitor
```

Comment:
Why the `Useful` AssertVisitor? Is there a `UselessAssertVisitor` at some place?

Reply:
Hehe, that was the old name (useful asserts explaining the crash instead of just crashing), I'll update it to ExplainAssertVisitor.